Iteration 13

mikesvis opened the pull request:
Хоть у меня и была обработка на создание дубликатов всю дорогу - переделал метод в соответствии с заданием.

mikesvis committed:
Iteration 10: Db storage + ping

mikesvis committed:
Iteration 11: DB implementation + context

mikesvis committed:
Iteration 12. Batch store aka Я_задолбался_с_дельтами

mikesvis committed:
Iteration 13: Added conflict error

capsulated commented on internal/api/api.go:
```diff
@@ -9,3 +17,55 @@ type Request struct {
+
+// А тут ли это должно быть?
+func (c *BatchRequest) UnmarshalJSON(data []byte) error {
+	type BatchRequestAlias BatchRequest
```
А алиас зачем?

mikesvis replied:
Упростил

capsulated commented on internal/api/api.go:
```diff
@@ -9,3 +17,55 @@ type Request struct {
+}
+
+// А тут ли это должно быть?
+func (c *BatchRequest) UnmarshalJSON(data []byte) error {
```
Я что-то даже не пойму, что это за метод))) Unmarshal должен быть где-то рядом основной с сущностью url'ов. Там должен появится, к примеру, слайс url'ов и можно добавить Marshal и Unmarshal методы для этих типов

mikesvis replied:
Упростил

capsulated commented on internal/api/api.go:
```diff
@@ -9,3 +17,55 @@ type Request struct {
+
+	tmp := &struct {
+		*BatchRequestAlias
+		CorrelationID string `json:"correlation_id"`
```
Я совсем не пойму, что тут присходит)) А correlation_id - это для чего?

mikesvis replied:
Упростил

capsulated replied:
correlation_id посмотрел в примерах в теории, это приемлемо

capsulated commented on internal/app/app.go:
```diff
@@ -23,7 +23,7 @@ type App struct {
 	config := config.NewConfig()
 	logger := logger.NewLogger()
-	storage := storage.NewStorage(string(config.FileStoragePath))
+	storage := storage.NewStorage(config)
```
👍

capsulated commented on internal/config/config.go:
```diff
@@ -51,11 +54,25 @@ func (s *FilePath) Type() string {
 	return "string"
 }
 
+func (d *DBAddress) Set(flagValue string) error {
```
Странный, всё же, этот тип для адреса. Я бы просто строки конкантенировал и всё )) Усложняют эти методы логику. Но смотри сам, это не критично

mikesvis replied:
Упростил

capsulated commented on internal/config/config.go:
```diff
@@ -51,11 +54,25 @@ func (s *FilePath) Type() string {
 }
 
+func (d *DBAddress) Set(flagValue string) error {
+	*d = DBAddress(string(flagValue))
```
Выглядит, грубо говоря, страшно ))

mikesvis replied:
Бес попутал, но упростил :)

capsulated replied:
Да не так уж и страшно-то. Создавать удобные для работы новые типы - это абсолютно правильный подход. Так как тип простой, то назначается с разименованием (на звонке 1-1 разобрались). Я тоже часто впорос просто задаю в комменте, если нет прямого призыва к действию - то можно просто ответить на вопрос))

capsulated commented on internal/drivers/filedb/filedb.go:
```diff
@@ -0,0 +1,183 @@
+}
+
+func (s *FileDB) StoreBatch(ctx context.Context, us map[string]domain.URL) (map[string]domain.URL, error) {
+	// в мапе хранится полный урл = ключ корреляции
```
А зачем его в "ключ корреляции" переименовывать? Пусть так и остаётся full url ) Тоже, на твоё усмотрение

mikesvis replied:
Это мапа  `map[fullUrl]correlation_id` так надол ))
Обсудили 1-1

capsulated commented on internal/drivers/memorymap/memorymap.go:
```diff
@@ -0,0 +1,96 @@
+	"github.com/mikesvis/short/internal/errors"
+)
+
+type MemoryMap struct {
```
`MemoryStore`, `MemoryStorage`, `MemoryRepo`, `MemoryRepository` или что-то подобное будет лучше. Указывать в имени переменной особенности реализации (в нашем случае - мапа) - это антипаттерн, так как    реализация легко может измениться, а тот же функционал может быть необходим. Вдруг возьмём и переедем на redis или какую-то стороннюю библиотеку хранилки в памяти

Обязательно надо переименовать по своему вкусу

mikesvis replied:
Переименовал на `InMemory`

capsulated commented on internal/drivers/memorymap/memorymap.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+func (s *MemoryMap) StoreBatch(ctx context.Context, us map[string]domain.URL) (map[string]domain.URL, error) {
+	// в мапе хранится полный урл = ключ корреляции
```
про ключ корреляции написал выше

mikesvis replied:
Угу, обсудили

capsulated commented on internal/drivers/postgres/postgres.go:
```diff
@@ -0,0 +1,220 @@
+	"github.com/mikesvis/short/internal/errors"
+)
+
+type postgresDBItem struct {
```
А эти item нельзя сделать едиными (глобальными) для всех хранилок разве? Если можно - то лучше так и сделать. Это вообще главная сущность нашего сервиса и она по идее, будет везде появляться почти во всех наших пакетах. Хороший вариант - создать глобальный пакет для этой сущности. Могут быть как единичные urlItem, так и множественные - []urlItem `type UrlItems []UrlItem` (неминг примерный). И в этом же пакете будут всякие маршалеры и анмаршеллеры для этих типов и тп

mikesvis replied:
Обсудили. Я не хочу перемешивать сущности базы с сущностями домена.

capsulated replied:
Ок ок

capsulated commented on internal/drivers/postgres/postgres.go:
```diff
@@ -0,0 +1,220 @@
+	return &Postgres{db}
+}
+
+func bootstrapDB(db *sql.DB) error {
```
Обычно, создаётся init.sql для контенера с базой. Но у нас с автотестами так не сделать) Это для спраки

mikesvis replied:
Угу, но пока так

capsulated replied:
Да, он так и останется)

capsulated commented on internal/drivers/postgres/postgres.go:
```diff
@@ -0,0 +1,220 @@
+}
+
+func (s *Postgres) findByColumn(ctx context.Context, column, needle string) (domain.URL, error) {
+	// вот это какой-то бред! как я не пытался передать в формирование запроса prepare колонку, не смог
```
Что-то я тоже не совсем понял, что тут за логика. Кажется, что можно упростить.

И да, выглядит как-то странно )))) И кажется, оно вообще  не надо) С виду, просто надо два разных метода сделать без switch с двумя разными селектами

mikesvis replied:
Сделал 2 метода и упростил, убрал Count

capsulated replied:
Да супер, в данном случае это намного лучше)

capsulated commented on internal/drivers/postgres/postgres.go:
```diff
@@ -0,0 +1,220 @@
+}
+
+func (s *Postgres) StoreBatch(ctx context.Context, us map[string]domain.URL) (map[string]domain.URL, error) {
+	// подход в рамках одной транзации: берем 1 элемент, смотрим есть ли он в базе,
```
А зачем транзакция? Есть специальные методы в библиотеке для БД, посмотри доку по внимательнее ne либe, которую используешь (в sqlx етсь спец batch запросы которые слайс могут записывать сразу, именнованные NamedExec вроде бы, в squirell специальная конструкция для подготовки таких запросов и тп). Также, есть "костыльный" способ - собрать строку (не желательный, но работающий). И select точно нужен? Было же выше использование sql конструкции `ON CONFLICT`. Также можно погуглить `INSERT OR UPDATE` / `UPSERT`. Вообще погуглить просто, batch INSERT с выбранными БД и библиотекой

mikesvis replied:
Опишу что обсудили на 1-1:
1) SELECT мне нужен для того чтобы узнать короткий сгенерированный url, который был добавлен в базу ранее (в рамках другого запроса например, другим пользователем или вчера/позавчера...)
2) В задании было сказано использовать ON CONFLICT. UPDATE в этом случае мне не подойдет, т.к. я сотру предыдущий короткий url, который был добавлен в базу ранее (в рамках другого запроса например, другим пользователем или вчера/позавчера...)
3) Пример реализации INSERT построчно я взял из курса, при этом такой подход описывался как приоритетный по скорости. Также если у нас в рамках одной транзакции INSERT одной строки - то при конфликте предыдущие строки (успешные/безконфликтные) будут добавлены. Если делать batch INSERT нескольких строк в рамках одной транзакции, то в случае конфликта транзакция откатится и не будет добавлен ни один. А что дальше? Выяснять какой конфликтует и вычеркивать его из добавления (поскольку он был добавлен "вчера") и снова пытаться batch INSERT? Тогда нужно "достать" shortUrl для конфликтных строк.

У меня же все проще:
В цикле по каждому элементу смотрим, есть ли он уже в базе по fullUrl. если он есть, читаем его shortKey (старый) и забываем про него. Если элемента нет, то INSERT.

В заданиях изначально не хватает условий, что делать если элемент уже был. Генерить дубль? Про дубли упоминается только в последней итерации в этом спринте:
```
Чтобы не проверять наличие оригинального URL в базе данных отдельным запросом, можно воспользоваться конструкцией INSERT ... ON CONFLICT в PostgreSQL. Однако в таком случае придётся самостоятельно возвращать и проверять собственную ошибку.
```
Вопрос: что делать по итогу? возвращать старый сгенерированный (что впринципе правильно), или же просто статус "Conflict". Тогда при добавленнии пользователей в сервис будет деградация урлов: кто первый добавил - тот сокращение и получит. Все последующие пользователи будут получать `Conflict` вместо короткого url (хоть и сгенерированного ранее).
Я же учитывал их с самого начала сервиса, поэтому у меня конфликта и быть не должно. НО! Раз в задании указали использовать `ON CONFLICT`, то я это переделал.

Я потратил очень много времени в предыдущих спринтах на поиск ранее сгенерированных shortUrl при добавлении fullUrl. Сейчас получается что я зря это делал?

mikesvis replied:
Я сделаль... Перевел на sqlx. Сделал batch insert. Это было самое сложное из всего.

capsulated replied:
Можно же ON CONFLICT DO NOTHING чтобы не презаписывать... Но нам надо будет вернуть же ещё 🤔

А sqlx - это ты лихо )) Но он реально очень удобный, ИМХО, по сравнению с остальными

capsulated commented on internal/drivers/postgres/postgres.go:
```diff
@@ -0,0 +1,220 @@
+	// мое мнение: это убогое решение, поскольку будет порождено столько транзакций/select'ов/insert'ов
+	// сколько элементов пришло в метод, реплике может стать очень плохо в перспективе кол-ва эл-тов на входе
+	// надо делать завдержку реплики
+	for k, v := range us {
```
Ну да, `for` и  запросы к БД - это плохое сочетание ))))

mikesvis replied:
Избавился от  сложности `O(N)`

capsulated replied:
А сложность от константы - это же неплохо. Вообще в цикле операции к БД или в цикле отправка http запросов - это явный признак, что надо упрощать)

capsulated commented on internal/server/handler.go:
```diff
@@ -29,8 +30,9 @@ func NewHandler(config *config.Config, storage storage.Storage) *Handler {
 // Получение короткого URL из запроса
 // Поиск в условной "базе" полного URL по сокращенному
 func (h *Handler) GetFullURL(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
```
Тут может дочерний создать: `ctx := context.With???(r.Context()...)`? Можно просто с отменой и defer cancle(). Стандартный подход (мало ли захотим сами выставить таймаут иди отменить для чего-то). Думаю не критично, но обычно создаются дочерние на уровне метода

mikesvis replied:
Вот тут я не подумал что можно отнаследоваться (сделать дочерний). Исправил.

capsulated replied:
Ну это тоже не критично же, скорее для справки

capsulated commented on internal/server/handler.go:
```diff
@@ -53,6 +55,7 @@ func (h *Handler) GetFullURL(w http.ResponseWriter, r *http.Request) {
 // Проверка на валидность URL
 // Запись сокращенного Url в условную "базу" если нет такого ключа
 func (h *Handler) CreateShortURLText(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
```
Да и зачем для него отдельную переменную, только память тратим (совсем немного конечно, но при большом rps будет чуть заметнее). Либо создавать дочерний, либо уже прям так и передавать (но это как-то это не по канонам )))

mikesvis replied:
Дыа

capsulated commented on internal/server/handler.go:
```diff
@@ -93,7 +95,7 @@ func (h *Handler) CreateShortURLText(w http.ResponseWriter, r *http.Request) {
 // Обработка всего остального
 func (h *Handler) Fail(w http.ResponseWriter, r *http.Request) {
-	err := errors.New("bad protocol")
+	err := ge.New("bad protocol")
```
`errors.New` понятно, а что за `ge`? Нейминг не очень удачный

mikesvis replied:
сделал `_goerrors`

capsulated replied:
это тоже минорное замечание. Тогда должны быть handlersErrors (уровня хендлера)

capsulated commented on internal/server/handler.go:
```diff
@@ -142,3 +145,66 @@ func (h *Handler) CreateShortURLJSON(w http.ResponseWriter, r *http.Request) {
+// Запись сокращенного URL в условную "базу" если нет такого ключа
+func (h *Handler) CreateShortURLBatch(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	var request []api.BatchRequest
```
По идее, надо прочитать байты body и преобразовать в слайс url'ов (к примеру, можно сделать спец конструктор для такого типа, с входным параметрам из байтов body). И если всё смаршаллилось без ошибок - слайс урлов передавать в метод хранилки и, в случае БД, batсh методом сохранить

mikesvis replied:
Это обсудили. Слайс не подойдет - нужна `map`.

capsulated replied:
Я что-то уже забыл, а почему слайс не подойдёт? Вообще слайсы структур - это же прям правильный объект для записи в БД. С sqlx вообще идеально. Хотя в sqlx можно и мапы писать в бд, вроде

capsulated commented on internal/server/handler.go:
```diff
@@ -142,3 +145,66 @@ func (h *Handler) CreateShortURLJSON(w http.ResponseWriter, r *http.Request) {
+		return
+	}
+
+	var response []api.BatchResponse
```
Вот сомневаюсь в доп типе для ответа, точно нужен? Мне кажется можно сделать два типа - единичная entity для url'а и слайс для множества. А далее уже преобразовывать, если необходимо. Не всегда это получается и не всегда это нужно, иногда локальные типы необходимы, нужно смотреть по ситуации. Но в идеале лучше минимизировать насколько можно количество типов, которые описывают одинаковые сущности

mikesvis replied:
Обсудили. Тут я все уростил, но структуры хочу оставить - не люблю я все в один котел, понимаю что в начале об этом думать не надо и лучше проще, но мне вероисповедание не позволяет завернуть все в одну структуру :) Слишком много я от такого в последствии избавлялся по опыту.

capsulated commented on internal/server/handler.go:
```diff
@@ -142,3 +145,66 @@ func (h *Handler) CreateShortURLJSON(w http.ResponseWriter, r *http.Request) {
+
+	// domain.URL.Short в процессе сохранения поменяем на старый если такой domain.URL.Full уже есть
+	// цель: сделать получение/вычисление/сохранение в одну транзакцию
+	// упарываться ответить кодом 201 если всех элементов не существовало ранее - не стал
```
ну если автотест пропускает - то мне особо это не важно ))

mikesvis replied:
Нужны требачи (Бизнес требования) Про дубли написал выше, жаль что с самого начала их учитывал и тратил время :( Можно было в 13 сделать и не париться, но опыт не пропьешь :)

capsulated replied:
У нас требования к автотесту. Если проверяет определённый статус - то ничего не поделаешь

capsulated commented on internal/storage/storage.go:
```diff
@@ -1,21 +1,46 @@
+	StoreBatch(ctx context.Context, pack map[string]domain.URL) (map[string]domain.URL, error)
+	GetByFull(ctx context.Context, fullURL string) (domain.URL, error)
+	GetByShort(ctx context.Context, shortURL string) (domain.URL, error)
+	Ping(ctx context.Context) error
```
Ох уж этот пинг, его же не должно быть у хранилки с памятью )))

mikesvis replied:
Уточните пожалуйста требования, ибо это уже вопрос к бизнесу (так как дальше есть варианты), а вопрос вот такой:
Как будет вести себя ручка `GET /ping` в случае хранилки типа `InMemory`? Какой статус, какой ответ, какая логика?

mikesvis replied:
А какие требачи?
Какой ответ должен быть если хранилка в памяти?
Внес доработку

capsulated replied:
Вообще поделить можно интерфейс. Но проверять нужно будет на соотвествие интерфесу

capsulated commented on internal/storage/storage.go:
```diff
@@ -1,21 +1,46 @@
+		if err != nil {
+			panic(err)
+		}
+		// defer db.Close() - вот с закрытием немного не ясно, поскольку тут закрытию точно не место
```
Всё верно, в конструкторе не должно быть. Это в `app` должно быть, но не у всех типов хранилок он есть) Решений много, но все не совсем лаконичные, будут выглядить 'костыльно'. Но может у тебя получится что-то изящное))) Может возвращать какие-то доп объекты в конструкторе (типа Pinger / Closer), и если они есть - вызывать. Либо приводить интерфейсный тип к конкретному `storage.(DbStorage).Ping()` -с добавлением проверки конфига или соответствия самого типа. Выглядит всё это не очень, нужно что-то изобрести ))) А в задании было реализовать ping, поэтому нужно сделать (для образовательных целей, по аналогии - такое тз пришло от заказчика и поменять его нельзя =)

mikesvis replied:
Внес доработку из серии `storage.(DbStorage).Ping()` с добавлением проверки типа

capsulated replied:
типа того, один из варинатов. Либо доп объекты - закрыватель, пингер или типа того, которые вызываются у нужных хранилок и игнорируются в других. Вариантов можнт быть много

capsulated reviewed (changes requested):
Коменты добавил. Если есть что-то спорное или непонятное - как обычно, можно писать в чат подгруппы или 1-1, если много вопросов

mikesvis committed:
Iteration 13 Corrections pt1

1) Closer, Pinger interfaces + implementation is drivers
2) Config FileStoragePath, DatabaseDSN as string
3) MemoryMap -> InMemory
4) InMemory removed Ping()
5) ge package naming as _goerrors
6) Postgres.GetByFull() / Postgres.GetByShort() simplyfied
7) Handler child context is extended from request context

mikesvis committed:
Iteration 13: Config fixes

mikesvis committed:
Iteration 13: Still config fixes

mikesvis committed:
Iteration 13: Final config fixes

mikesvis committed:
Iteration 13: Simpyfying Batch Request / Response

mikesvis committed:
Iteration 13: Handler fixes, Store() simplification

mikesvis committed:
Iteration 13: SQLX in batch

capsulated reviewed (approved):
Большая работа проделана, много комментариев и исправлений. Есть большой плюс в этом - ты быстрее привыкаешь с синтаксису и структуре программы на го

Апрувлю

А в следующих спринтах уже будем смотреть, в процессе дальнейшей разработки, надо что-то переорганизовывать для удобства или нет
